Admin user and jenkins new scheme

pcm32 opened the pull request:
This PR:

- Allows automatic setup of a master api key, which is injected through helm.
- Allows injection of user admin email and password as ENV variables, to setup initial admin user.
- Removes master api key once user is set.
- Adds necessary simple tests and annotations to docker file to adhere to the new Jenkins job setup.

ilveroluca commented on Dockerfile:
```diff
@@ -1,6 +1,10 @@
+LABEL Description="Galaxy 16.07-phenomenal for running inside Kubernetes."
+LABEL software="Galaxy"
+LABEL software.version="16.07-pheno"
+LABEL version="0.1"
```
These LABEL statements could be compressed into a single one, reducing the number of layers in the container.

pcm32 replied:
We use these labels to be in accordance to what biocontainers recommends. We need them as well for our versioning setup in Jenkins.

ilveroluca replied:
Sorry, I was ambiguous :-)  I meant using a single LABEL command, not a single label.  From what I understand this should generate a single layer and set the same labels:
```
LABEL Description="Galaxy 16.07-phenomenal for running inside Kubernetes."  \
      software="Galaxy" \
      software.version="16.07-pheno" \
      version="0.1"
```

pcm32 replied:
Good one! But my impression is that Pierrick's code at the other side on Jenkins is expecting the "^LABEL (key)=(value)" regexp, so I would leave it like that on this one.

ilveroluca commented on ansible/create_admin_user.py:
```diff
@@ -0,0 +1,32 @@
+gi = bioblend.galaxy.GalaxyInstance(galaxy_url, galaxy_api_key)
+
+# Create a new user
+new_user = gi.users.create_local_user(username, email, password)
```
Missing error check.  Call may fail silently.

pcm32 replied:
Agreed! Saw this fail silently many times!! How would you check for an error in this case? My experience with bioblend is reduced to, well, those lines :-).

ilveroluca replied:
Bioblend docs don't say.  I'll look into it right now...

ilveroluca replied:
So, I did a few tests with Bioblend 0.8.0 and it appears the `create_local_user` call shouldn't fail silently.

I found three cases.

1. Fails to connect: raises a request.RequestException -> IOError.  The exception type isn't documented so it's probably safest to catch `StandardError`.
2. Connects but Galaxy gives an error: raises a bioblend.ConnectionError
3. Success:  returns a dictionary with some info about the user.  Here's an example
```{u'email': u'pireddu@crs4.it',
 u'id': u'f597429621d6eb2b',
 u'model_class': u'User',
 u'nice_total_disk_usage': u'0 bytes',
 u'total_disk_usage': 0.0,
 u'username': u'luca'}
```
Rather than depending on the exceptions, we could just print `new_user['id']`.  If bioblend doesn't raise we'll at least have the access to the dictionary raise an exception.

ilveroluca commented on ansible/run_galaxy_config.sh:
```diff
@@ -3,3 +3,38 @@ sudo apt-get update -y && sudo apt-get install -y --no-install-recommends ansibl
+	# ini file will be set already at this point with these variables, we only need to start galaxy
+	# wait for it to be available for API calls and create the new user
+	echo "Set up galaxy database for admin user setup..."
+	./run.sh --daemon
```
It would be defensive to `rm -f paster.pid` before calling `run.sh` to avoid looping forever if the file ever doesn't get written properly (not sure whether this error case is currently possible).

pcm32 replied:
Never saw it happening on my tests, but we can add it certainly.

pcm32 replied:
won't the script fail if we try to rm and the file is not there, given that we set `set -o errexit`?

ilveroluca replied:
`rm -f ` should always return 0, even if the file doesn't exist.

pcm32 replied:
Good, will add both, thanks.

ilveroluca commented on ansible/run_galaxy_config.sh:
```diff
@@ -3,3 +3,38 @@ sudo apt-get update -y && sudo apt-get install -y --no-install-recommends ansibl
+	echo "Galaxy is ready for API calls..."
+	echo "Running admin user creation..."
+	source .venv/bin/activate
+	python ansible/create_admin_user.py $GALAXY_API_KEY $GALAXY_ADMIN_USER $GALAXY_ADMIN_EMAIL $GALAXY_ADMIN_PASSWORD
```
Check exit code or we risk a silent failure.

I like to use `set -o errexit` and `set -o nounset` in my bash scripts to take care of this automatically.

pcm32 commented:
I think I have addressed the issues now. I also added a check to avoid trying to create the user again if it already exists. Unfortunately this needs to happen through the API and with the master_api_key set, so still getting into the loop of restarting Galaxy.
